### PR TITLE
Improve the KATalogus `/plugins` endpoint performance

### PR DESCRIPTION
**Pull request by Donnype**
### Changes

And add a bulk endpoint for the declarations of the CSV upload. The main fixes in the KATalogus are:
- Do not query the database per boefje id on each request to fetch the `enabled` state, just do 1 query per request
- Query both boefje and normalizers at the same time
- Do not query the existence of the organization on each request: this is added as a filter parameter anyways and this should be checked in Rocky
- Cache boefje/normalizer module loading not on an instance-level but with a function @cache decorator, since then we have the cache available for other requests as well where before we had a new instance of the local repository per request.

### Issue link

Closes https://github.com/minvws/nl-kat-coordination/issues/3903
Closes https://github.com/minvws/nl-kat-coordination/issues/2524

#### Before

Snapshot of `htop` (also notice the postgres process because we did 100+ queries per second, because we managed to handle 2 requests to the `/plugins` endpoint per second):
Scherm­afbeelding 2024-12-02 om 09 54 43

Requests per second (RPS) of the `plugin` endpoint, tested using [Locust](https://locust.io/):
Scherm­afbeelding 2024-11-29 om 11 34 50

#### After
Requests per second (RPS) of the `plugin` endpoint, tested using [Locust](https://locust.io/):
Scherm­afbeelding 2024-11-29 om 11 32 31

Snapshot of `htop` (the boefje api processes where nowhere near the top):
Scherm­afbeelding 2024-12-02 om 08 30 43

### QA notes

Please verify that the CSV upload still works and the scheduler picks up tasks, also from duplicated boefjes! Perhaps use the [test server](https://kattest03.openkat.dev/) that I used to verify the interface works OK with 300 organizations.

---

### Code Checklist

- [x] All the commits in this PR are properly PGP-signed and verified.
- [x] This PR only contains functionality relevant to the issue.
- [x] I have written unit tests for the changes or fixes I made.
- [x] I have checked the documentation and made changes where necessary.
- [x] I have performed a self-review of my code and refactored it to the best of my abilities.

---

## Checklist for code reviewers:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---

## Checklist for QA:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.

### CODE REVIEW

**Comment by ammar92** [comment]
> * Do not query the database per boefje id on each request to fetch the `enabled` state, just do 1 query per request

Cool, this essentially fixes https://github.com/minvws/nl-kat-coordination/issues/2524 right?

File: boefjes/boefjes/local_repository.py [hunk 1]
```diff
@@ -107,66 +103,90 @@ def description_path(self, id_: str) -> Path | None:
...
 
-        for path, package in paths_and_packages:
-            try:
-                boefje_resources.append(BoefjeResource(path, package))
-            except ModuleException as exc:
-                logger.exception(exc)
 
-        self._cached_boefjes = {resource.boefje.id: resource for resource in boefje_resources}
+@cache
+def _cached_resolve_boefjes(path) -> dict[str, BoefjeResource]:
```
Comment (dekkers) [inline boefjes/boefjes/local_repository.py:113]:
```suggestion
def _cached_resolve_boefjes(path: Path) -> dict[str, BoefjeResource]:
```

File: boefjes/boefjes/local_repository.py [hunk 2]
```diff
@@ -107,66 +103,90 @@ def description_path(self, id_: str) -> Path | None:
...
+    """Adding the hash to the arguments makes sure we refresh this. The size could hence be 1, but since this is not
```
Comment (dekkers) [inline boefjes/boefjes/local_repository.py:114]: There isn't a hash in the arguments, only the path, was the docstring was written for a previous version?

**Review by dekkers — COMMENTED** [review]
Looks good, nice work :+1:

File: boefjes/boefjes/local_repository.py (hunk 2 above)
Comment (Donnype) [reply boefjes/boefjes/local_repository.py:114]: Indeed!

**Comment by jpbruinsslot** [comment]
Tested it locally, added 300 organisations and added 1000 hostnames set to clearance 1. Didn't notice any irregularities

File: boefjes/boefjes/local_repository.py [hunk 3]
```diff
@@ -107,66 +103,82 @@ def description_path(self, id_: str) -> Path | None:
...
+    for path, package in paths_and_packages:
+        try:
+            normalizer_resources.append(get_normalizer_resource(path, package, hash(path)))
+        except ModuleException as exc:
+            logger.exception(exc)
+
+    return {resource.normalizer.id: resource for resource in normalizer_resources}
+
+
+def _find_packages_in_path_containing_files(path, required_files: tuple[str, ...]) -> list[tuple[Path, str]]:
```
Comment (ammar92) [inline boefjes/boefjes/local_repository.py:142]:
```suggestion
def _find_packages_in_path_containing_files(path: Path, required_files: tuple[str, ...]) -> list[tuple[Path, str]]:
```

File: boefjes/boefjes/local_repository.py [hunk 4]
```diff
@@ -107,66 +103,82 @@ def description_path(self, id_: str) -> Path | None:
...
-            missing_files = [file for file in required_files if not (path / file).exists()]
+    return f"{relative_path[1:].replace('/', '.')}."  # Turns into "boefjes.plugins."
 
-            if missing_files:
-                logger.debug("Files %s not found for %s", missing_files, package.name)
-                continue
 
-            paths.append((path, package.name))
+@lru_cache(maxsize=200)
+def get_boefje_resource(path, package, path_hash):
```
Comment (ammar92) [inline boefjes/boefjes/local_repository.py:170]:
```suggestion
def get_boefje_resource(path: Path, package: str, path_hash: str) -> BoefjeResource:
```

File: boefjes/boefjes/local_repository.py [hunk 5]
```diff
@@ -107,66 +103,82 @@ def description_path(self, id_: str) -> Path | None:
...
+    space."""
+    return BoefjeResource(path, package, path_hash)
 
-        return paths
 
-    @staticmethod
-    def create_relative_import_statement_from_cwd(package_dir: Path) -> str:
-        relative_path = str(package_dir.absolute()).replace(str(Path.cwd()), "")  # e.g. "/boefjes/plugins"
+@lru_cache(maxsize=200)
+def get_normalizer_resource(path, package, path_hash):
```
Comment (ammar92) [inline boefjes/boefjes/local_repository.py:177]:
```suggestion
def get_normalizer_resource((path: Path, package: str, path_hash: str) -> NormalizerResource:
```

**Review by ammar92 — COMMENTED** [review]
Nice work. Tiny suggestions, consider it approved

**Review by underdarknl — APPROVED** [review]

**Pull request merged**